backends-common/rabbitmq/src/test/java/org/apache/james/backend/rabbitmq/RabbitMQExtension.java

```diff
@@ -32,19 +34,108 @@
-    private DockerRabbitMQ rabbitMQ;
+
+    private static final Consumer<DockerRabbitMQ> DO_NOT_THING = dockerRabbitMQ -> {};
+
+    public enum DockerRestartPolicy {
+        PER_TEST(DockerRabbitMQ::start, DockerRabbitMQ::start, DockerRabbitMQ::stop, DockerRabbitMQ::stop),
+        PER_CLASS(DockerRabbitMQ::start, DO_NOT_THING, DO_NOT_THING, DockerRabbitMQ::stop),
+        NEVER(DockerRabbitMQ::start, DO_NOT_THING, DO_NOT_THING, DO_NOT_THING);
+
+        private final Consumer<DockerRabbitMQ> beforeAllCallback;
+        private final Consumer<DockerRabbitMQ> beforeEachCallback;
+        private final Consumer<DockerRabbitMQ> afterEachCallback;
+        private final Consumer<DockerRabbitMQ> afterAllCallback;
+
+        DockerRestartPolicy(Consumer<DockerRabbitMQ> beforeAllCallback,
+                            Consumer<DockerRabbitMQ> beforeEachCallback,
+                            Consumer<DockerRabbitMQ> afterEachCallback,
+                            Consumer<DockerRabbitMQ> afterAllCallback) {
+            this.beforeAllCallback = beforeAllCallback;
+            this.beforeEachCallback = beforeEachCallback;
+            this.afterEachCallback = afterEachCallback;
+            this.afterAllCallback = afterAllCallback;
+        }
+
+        public void beforeAll(DockerRabbitMQ dockerRabbitMQ) {
+            beforeAllCallback.accept(dockerRabbitMQ);
+        }
+
+        public void afterAll(DockerRabbitMQ dockerRabbitMQ) {
+            afterAllCallback.accept(dockerRabbitMQ);
+        }
+
+        public void afterEach(DockerRabbitMQ dockerRabbitMQ) {
+            afterEachCallback.accept(dockerRabbitMQ);
+        }
+
+        public void beforeEach(DockerRabbitMQ dockerRabbitMQ) {
+            beforeEachCallback.accept(dockerRabbitMQ);
+        }
+    }
+
+    public static class Builder {
+
+        @FunctionalInterface
+        public interface RequireDockerRabbitMQ {
+            RequireRestartPolicy dockerRabbitMQ(DockerRabbitMQ dockerRabbitMQ);
+
+            default RequireRestartPolicy defaultRabbitMQ() {
+                return dockerRabbitMQ(DockerRabbitMQ.withoutCookie());
+            }
+
+            default ReadyToBuild singletonRabbitMQ() {
+                return new ReadyToBuild(DockerRabbitMQSingleton.SINGLETON, NEVER);
+            }
+        }
+
+        @FunctionalInterface
+        public interface RequireRestartPolicy {
+            ReadyToBuild restartPolicy(DockerRestartPolicy dockerRestartPolicy);
+        }
+
+        public static class ReadyToBuild {
+
+            private final DockerRabbitMQ rabbitMQ;
+            private final DockerRestartPolicy dockerRestartPolicy;
+
+            private ReadyToBuild(DockerRabbitMQ rabbitMQ, DockerRestartPolicy dockerRestartPolicy) {
+                this.rabbitMQ = rabbitMQ;
+                this.dockerRestartPolicy = dockerRestartPolicy;
+            }
+
+            public RabbitMQExtension build() {
```

Comment:
IMO the build step is useless in this case, we can build the extension when we choose the policy